// @baklavajs/core/dist/cjs/nodeInterface.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.NodeInterface = void 0;
const uuid_1 = require("uuid");
const events_1 = require("@baklavajs/events");
class NodeInterface {
    set connectionCount(v) {
        this._connectionCount = v;
        this.events.setConnectionCount.emit(v);
    }
    get connectionCount() {
        return this._connectionCount;
    }
    set value(v) {
        if (this.events.beforeSetValue.emit(v).prevented) {
            return;
        }
        this._value = v;
        this.events.setValue.emit(v);
    }
    get value() {
        return this._value;
    }
    constructor(name, value) {
        this.id = (0, uuid_1.v4)();
        /** Id of the node this interface is part of */
        this.nodeId = "";
        /** Whether to show the port (the thing connections connect to) */
        this.port = true;
        /** Whether to hide the node interface in the UI */
        this.hidden = false;
        this.events = {
            setConnectionCount: new events_1.BaklavaEvent(this),
            beforeSetValue: new events_1.PreventableBaklavaEvent(this),
            setValue: new events_1.BaklavaEvent(this),
            updated: new events_1.BaklavaEvent(this),
        };
        this.hooks = {
            load: new events_1.SequentialHook(this),
            save: new events_1.SequentialHook(this),
        };
        this._connectionCount = 0;
        this.name = name;
        this._value = value;
    }
    load(state) {
        this.id = state.id;
        this.templateId = state.templateId;
        this.value = state.value;
        this.hooks.load.execute(state);
    }
    save() {
        const state = {
            id: this.id,
            templateId: this.templateId,
            value: this.value,
        };
        return this.hooks.save.execute(state);
    }
    setComponent(value) {
        this.component = value;
        return this;
    }
    setPort(value) {
        this.port = value;
        return this;
    }
    setHidden(value) {
        this.hidden = value;
        return this;
    }
    use(middleware, ...args) {
        middleware(this, ...args);
        return this;
    }
}
exports.NodeInterface = NodeInterface;
